arch/risc-v/include/mode.h

```diff
@@ -0,0 +1,101 @@
+/****************************************************************************
+ * arch/risc-v/include/mode.h
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __ARCH_RISCV_INCLUDE_MODE_H
+#define __ARCH_RISCV_INCLUDE_MODE_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/irq.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#ifdef CONFIG_ARCH_USE_S_MODE
+
+/* CSR definitions */
+
+#  define CSR_STATUS        sstatus          /* Global status register */
```

Comment:
Can we

```suggestion
#  define CSR_STATUS        __XSTR(sstatus)          /* Global status register */
```
And remove `__XSTR` from other places. That was my main idea when I was adding those macro generation logic

Reply:
Makes sense, this way the macros are usable from asm and inline asm

Reply:
Still the issue remains, including riscv_internal.h into this file will end in disaster

Reply:
I do not understand why. You moved the logic to `arch/risc-v/include/macros.h`.
Let me check your change locally and try to see what "will end in disaster" means. Could you please send me steps to reproduce it?

Reply:
My point was that I don't think #include "riscv_internal.h" will work for this file. This is the reason I created the new file, so it can be included from /src and /include folders. I'm not 100% familiar with how the -I flags are provided but if I remember correctly, including something from the /src folder works for files that are in the /src folder or higher (relatively)

Reply:
Initially I thought to place macro in arch/risc-v/include/arch.h but didn't found any strong reason to do that in my initial change.

Reply:
I still think that `arch/risc-v/include/arch.h` is a good place for it

Reply:
you can either switch to `arch/risc-v/include/arch.h` or I can do it to minimize changes in this PR

Reply:
let me do the change and get wider visibility for macro

Reply:
After doing the suggestion, the WRITE_CSR() etc macros stop working:
chip/mpfs_irq.c:135: Error: unknown CSR `sie'
chip/mpfs_irq.c:135: Error: unknown CSR `sie'
chip/mpfs_irq.c:141: Error: unknown CSR `sie'
chip/mpfs_irq.c:178: Error: unknown CSR `sie'
chip/mpfs_irq.c:178: Error: unknown CSR `sie'
chip/mpfs_irq.c:184: Error: unknown CSR `sie'
chip/mpfs_irq.c:231: Error: unknown CSR `sie'
chip/mpfs_irq.c:235: Error: unknown CSR `sstatus'
chip/mpfs_irq.c:235: Error: unknown CSR `sstatus'

#  define CSR_IE            __XSTR(sie)      /* Interrupt enable register */
CLEAR_CSR(CSR_IE, IE_SIE);

#define WRITE_CSR(reg, val) \
  ({ \
     __asm__ __volatile__("csrw " __STR(reg) ", %0" :: "rK"(val)); \
  })

I have no idea how the substitution works in this case

Reply:
let me check

Reply:
> let me do the change and get wider visibility for macro

By all means. I can live locally with my modifications and work on something else.

However, at this point I would like to share that I have ~20 unpublished commits related to this and kernel mode in my local repository. I spent the entire day fixing merge issues (every day this week in fact), and this will happen perpetually until I can finally get some of my work merged. So my wish is to not defer this PR unnecessarily any longer.

Reply:
May I suggest that we merge an implementation which works for @pussuw , even if there is room for improvement. @pkara can then beautify things in another pr if needed/wanted? We do not need to solve every possible style issue in one PR. This has already ~300 comments an frankly I see it becoming unbearable for the author.

Reply:
> I have no idea how the substitution works in this case

The solution is obviously to remove the double substitution, but this means that all other platforms need to be moved to use the mode dependent macros. The rabbit hole never ends, so at some point an intermediate stop would be much appreciated. My intent is not to shove everything in and then forget it. I take responsibility for what I do.

Reply:
Sure. Let's merge something that works.

Reply:
@pussuw what about `arch/risc-v/include/arch.h`? Or you think that `macros.h` is fine?

Reply:
The rebase is trivial. I will push rebased version to my version of nuttx and open a PR to your github for visibility.

Reply:
https://github.com/tiiuae/nuttx/pull/38

Reply:
> [tiiuae#38](https://github.com/tiiuae/nuttx/pull/38)

The faulty code is not yet published, it was a local error. The #5922 patch is fine.
As I said I have ~20 commits ready locally, so any (potential) merge errors they contain are not fixable by you :)

Reply:
Anyway I do what I can :) please tell me how I can help

Reply:
> Anyway I do what I can :) please tell me how I can help

Yes, much appreciated.

Reply:
I just re-read this thread again and think that both authors and reviewers are a bit tired and it is good not to rock the boat and stabilize this PR to get it merged.

When I was writing a comment to add `__XSTR()` I was expecting that other macros will be adapted accordingly like:
```
# define CSR_IE __XSTR(sie) /* Interrupt enable register */
CLEAR_CSR(CSR_IE, IE_SIE);

#define WRITE_CSR(reg, val) \
  ({ \ __asm__ __volatile__("csrw " reg ", %0" :: "rK"(val)); \ }) 

```
And stringify is removed from `WRITE_CSR` definition.

Reply:
> I just re-read this thread again and think that both authors and reviewers are a bit tired and it is good not to rock the boat and stabilize this PR to get it merged.
> 
> When I was writing a comment to add `__XSTR()` I was expecting that other macros will be adapted accordingly like:
> 
> ```
> # define CSR_IE __XSTR(sie) /* Interrupt enable register */
> CLEAR_CSR(CSR_IE, IE_SIE);
> 
> #define WRITE_CSR(reg, val) \
>   ({ \
>     __asm__ __volatile__("csrw " reg ", %0" :: "rK"(val)); \
>   }) 
> ```
> 
> And stringify is removed from `WRITE_CSR` definition.
> 
> EDIT 1: But with this approach any `reg` parameter should go thru `__XSTR()` in prior to be passed and that is not user friendly. Let's keep the things like they are now.

Correct, the parameters to WRITE_CSR() and others should be given by the mode.h macros on all platforms, but that fix can be done later.